server/src/main/java/org/elasticsearch/search/lookup/StoredFieldSourceProvider.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.lookup;
+
+import org.apache.lucene.index.CompositeReaderContext;
+import org.apache.lucene.index.LeafReaderContext;
+import org.elasticsearch.index.fieldvisitor.LeafStoredFieldLoader;
+import org.elasticsearch.index.fieldvisitor.StoredFieldLoader;
+
+import java.io.IOException;
+
+class StoredFieldSourceProvider implements SourceProvider {
+
+    private final StoredFieldLoader storedFieldLoader;
+    private volatile LeafStoredFieldSourceProvider[] leaves;
+
+    StoredFieldSourceProvider(StoredFieldLoader storedFieldLoader) {
+        this.storedFieldLoader = storedFieldLoader;
+    }
+
+    @Override
+    public Source getSource(LeafReaderContext ctx, int doc) throws IOException {
+        LeafStoredFieldSourceProvider[] leaves = getLeavesUnderLock(ctx.parent);
+        if (leaves[ctx.ord] == null) {
+            // individual segments are only accessed on one thread so there's no need
```

Comment:
This is true as far as concurrency is done at segment level. If/When inter segment concurrency is implemented, then we might need extra locking here? If this is true we might want to add it in the comment.

Reply:
If we implement within-segment concurrency then this will have to work completely differently I think.  I'll add a comment though.